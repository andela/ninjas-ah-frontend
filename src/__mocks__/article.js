export default (Object.article = {
  title: 'my title well done',
  description: 'description of the article: well',
  body: 'body of the article: well'
});
