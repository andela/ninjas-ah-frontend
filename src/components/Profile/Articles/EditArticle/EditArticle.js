import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import TextareaAutosize from 'react-autosize-textarea';
import MetaTags from 'react-meta-tags';
import { Link } from 'react-router-dom';
import { EditorState, convertFromRaw, convertToRaw } from 'draft-js';
import { Editor } from 'react-draft-wysiwyg';
import 'react-draft-wysiwyg/dist/react-draft-wysiwyg.css';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faQuestionCircle, faPen } from '@fortawesome/free-solid-svg-icons';
import { Alert, NotFound } from '../../../common';
import { editPost, fetchOneArticle } from '../../../../actions';
import Layout from '../../../Layout';
import './EditArticle.scss';
import Heading from '../../../common/Heading/Heading';
import timeStamp from '../../../../helpers/timeStamp';
import richTextConfig from '../../../../helpers/richTextConfig.json';

export class EditArticle extends Component {
  state = {
    pageTitle: 'Edit a story',
    title: '',
    description: '',
    body: '',
    article: '',
    slug: '',
    message: '',
    status: '',
    coverUrl: null,
    loaded: false,
    displayUploadButton: false,
    editorState: EditorState.createEmpty(),
    location: ''
  };

  async componentDidMount() {
    const { match: { params: { slug } } } = this.props;
    this.props.fetchOneArticle(slug).then(() => {
      this.setState({
        message: '',
        loaded: true,
        title: this.props.article.title,
        description: this.props.article.description,
        body: this.props.article.body,
        article: this.props.article,
        status: this.props.article.status
      });
      this.convertToEditorState(this.state.body);
    });
    this.setState({ slug });
  }

  convertToEditorState = (articleBody = '{}') => {
    const editor = convertFromRaw(JSON.parse(articleBody));
    const editorState = EditorState.createWithContent(editor);
    this.setState({ editorState });
  };

  componentWillReceiveProps(nextProps) {
    if (nextProps.article) {
      this.setState({ article: nextProps.article });
      this.setState({ message: nextProps.message });
    }
    if (nextProps.errors) {
      this.setState({ errors: nextProps.errors });
    }
  }

  onChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  };

  onEditorStateChange = editorState => this.setState({ editorState });

  onSubmit = async (e) => {
    const { slug } = this.state;
    e.preventDefault();
    const post = {
      title: this.state.title,
      description: this.state.description,
      body: JSON.stringify(convertToRaw(this.state.editorState.getCurrentContent()))
    };
    await this.props.editPost({ article: post, slug });
    this.setState({ article: this.props.article });
  };

  render() {
    const { errors } = this.props;
    const {
      pageTitle,
      title,
      description,
      message,
      article,
      editorState,
      status,
      loaded
    } = this.state;

    return (
      <Layout>
        <div id="editArticle">
          <div className="container">
            {errors
              && (errors.errors || []).map(error => (
                <div className="text-danger border b-light medium-padding text-white" key={error}>
                  <FontAwesomeIcon icon={faQuestionCircle} /> {error}
                </div>
              ))}
            {article && Object.keys(article).length > 0 && !errors.error ? (
              <div className="row">
                <MetaTags>
                  <title>Editing: {title || article.title} - Authors Haven </title>
                  <meta
                    name="description"
                    content="Create a community of like minded authors to foster inspiration and innovation by leveraging the modern web"
                  />
                </MetaTags>
                <div className="small-screen-4 medium-screen-3 large-screen-3">
                  <form onSubmit={this.onSubmit}>
                    {Object.keys(message).length ? (
                      <Alert alertType="success" message={message.message} />
                    ) : (
                      ''
                    )}
                    <div className="row">
                      <h2 className="medium-h-padding">{pageTitle}</h2>
                      <div className="input-field">
                        <TextareaAutosize
                          name="title"
                          onChange={this.onChange}
                          value={title}
                          placeholder="Type title here..."
                          style={{ minHeight: 15 }}
                          id="articleTitle"
                        />
                      </div>
                      <div className="input-field">
                        <TextareaAutosize
                          name="description"
                          onChange={this.onChange}
                          value={description}
                          placeholder="Story description..."
                          style={{ minHeight: 15 }}
                          id="articleDescription"
                        />
                      </div>
                      <div className="input-field">
                        <br />
                        <Editor
                          name="body"
                          editorState={editorState}
                          toolbarClassName="toolbarClassName"
                          wrapperClassName="wrapperClassName"
                          editorClassName="editorClassName"
                          onEditorStateChange={this.onEditorStateChange}
                          toolbar={richTextConfig}
                        />
                      </div>
                    </div>
                    <button
                      type="submit"
                      id="submit"
                      className="button radius-4 primary text-black bold"
                    >
                      Update
                    </button>
                    <br />
                    {Object.keys(errors).length && errors.message ? (
                      <div className="medium-padding medium-v-margin border b-light text-danger">
                        Sorry, error occurred. try again later
                      </div>
                    ) : (
                      ''
                    )}
                  </form>
                </div>
                <div className="small-screen-4 medium-screen-3 large-screen-1">
                  <div className="box article-actions">
                    <Link
                      to={`/profile/article/preview/${article.slug}`}
                      className="button block success text-white center-align radius-4"
                    >
                      <FontAwesomeIcon icon={faPen} /> Preview
                    </Link>
                  </div>
                  <div className="card">
                    <Heading type={1} text={'Details'} />
                    <div className="box">
                      <div className="medium-padding">
                        <div className="medium-v-padding">
                          Status: <span className="bold">{status || article.status}</span>
                        </div>

                        <div className="medium-v-padding">
                          Created:
                          <span>{timeStamp(article.updatedAt)}</span>
                        </div>
                        <div className="medium-v-padding">
                          Updated: <span>{timeStamp(article.updatedAt)}</span>
                        </div>
                        <div className="divider light" />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            ) : (
              <div>{loaded && !Object.keys(article).length ? <NotFound /> : <div>{''}</div>}</div>
            )}
          </div>
        </div>
        <div className="divider light" />
      </Layout>
    );
  }
}

EditArticle.defaultProps = { match: { params: { slug: '' } } };

EditArticle.propTypes = {
  editPost: PropTypes.func.isRequired,
  article: PropTypes.object,
  fetchOneArticle: PropTypes.func.isRequired,
  slug: PropTypes.string,
  match: PropTypes.object,
  params: PropTypes.object,
  message: PropTypes.object,
  errors: PropTypes.object,
  content: PropTypes.string
};
export const mapStateToProps = ({ articles: { article, message, errors } }) => ({
  article,
  message,
  errors
});

export default connect(
  mapStateToProps,
  { editPost, fetchOneArticle }
)(EditArticle);
