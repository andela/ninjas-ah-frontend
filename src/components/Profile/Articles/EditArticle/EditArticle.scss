@import '../../../../assets/css/_padding';
@import '../../../../assets/css/_colors';
@import '../../../../assets/css/_shadow';
#editArticle,
#createArticle {
  textarea[name='title'] {
    border: 0 !important;
    font-size: 40px;
    padding: 0 !important;
  }
  textarea[name='description'] {
    border: 0 !important;
    font-size: 30px;
    padding: 0 !important;
  }
  textarea[name='body'] {
    border: 0 !important;
    font-size: 20px;
    padding: 0 !important;
  }
  .input-field input:focus,
  .input-field textarea:focus {
    outline-width: 0;
    outline: none !important;
    box-shadow: none !important;
    -moz-box-shadow: none !important;
    -webkit-box-shadow: none !important;
  }
  .box {
    margin-bottom: $medium;
    border: 1px solid $light;
    padding: $small/2;
    @extend .shadow-1;
  }
  .tags-response {
    margin-left: 10px;
    margin-right: 10px;
    line-height: 15px;
  }
}
