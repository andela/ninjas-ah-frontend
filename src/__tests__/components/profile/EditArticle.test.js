import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'react-thunk';
import { Provider } from 'react-redux';
import article from '../../../__mocks__/article';
import { EditArticle as EditArticleComponent } from '../../../components/Profile/Articles/EditArticle/EditArticle';
import { shallow, mount } from '../../../../config/enzymeConfig';

const props = {
  fetchOneArticle: jest.fn(async () => {
    await true;
  }),
  errors: { errors: ['article not found'] },
  article,
  message: { message: 'article edited successfully' },
  match: { params: { slug: 'slug-slug-slug' } },
  editPost: jest.fn(),
  onEditorStateChange: jest.fn()
};
const state = {
  article: { title: 'hello', description: 'hello' },
  message: 'hello',
  title: 'hello',
  slug: 'slug-slug-slug',
  body: JSON.stringify({
    blocks: [
      {
        key: 'cnu26',
        text: 'test componentWillReceiveProps failedtest componentWillReceiveProps failed',
        type: 'unstyled',
        depth: 0,
        inlineStyleRanges: [
          { offset: 0, length: 74, style: 'color-rgb(36,41,46)' },
          { offset: 0, length: 74, style: 'bgcolor-rgb(255,255,255)' },
          { offset: 0, length: 74, style: 'fontsize-32' },
          {
            offset: 0,
            length: 74,
            style:
              'fontfamily--apple-system, system-ui, "Segoe UI", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol'
          }
        ],
        entityRanges: [],
        data: { 'text-align': 'start' }
      },
      {
        key: 'emuik',
        text: 'Okey',
        type: 'unstyled',
        depth: 0,
        inlineStyleRanges: [
          { offset: 0, length: 4, style: 'color-rgb(36,41,46)' },
          { offset: 0, length: 4, style: 'bgcolor-rgb(255,255,255)' },
          { offset: 0, length: 4, style: 'fontsize-32' },
          {
            offset: 0,
            length: 4,
            style:
              'fontfamily--apple-system, system-ui, "Segoe UI", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol'
          }
        ],
        entityRanges: [],
        data: {}
      }
    ],
    entityMap: {}
  })
};
const component = shallow(<EditArticleComponent {...props} />);
describe('<EditArticleComponent />', () => {
  it('should render a <EditArticleComponent /> component ', () => {
    component.setProps({ article: state.article });
    expect(component).toMatchSnapshot();
    component.instance().convertToEditorState(state.body);
  });
  it('should create articles', () => {
    const instance = component.instance();
    const fakeEvent = { preventDefault: () => {} };
    instance.onSubmit(fakeEvent);
  });
  it('should call onChange method when the body value is changed', () => {
    const instance = component.instance();
    component.setProps({ ...props });
    instance.onEditorStateChange();
    expect(instance).toBeDefined();
  });
  it('should call onChange method when the description value is changed', () => {
    const spy = jest.spyOn(component.instance(), 'onChange');
    component.instance().forceUpdate();
    component.setProps({ ...props, errors: {} });
    const event = { target: { value: 'description' } };
    const textArea = component.find('#articleDescription');
    textArea.simulate('change', event);
    expect(textArea.length).toBe(1);
    expect(spy).toHaveBeenCalled();
  });
  it('should create articles', () => {
    const instance = component.instance();
    const fakeEvent = { preventDefault: () => {} };
    instance.onSubmit(fakeEvent);
  });
});
