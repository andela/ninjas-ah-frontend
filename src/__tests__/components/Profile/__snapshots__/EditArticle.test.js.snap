// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditArticleComponent /> should render a <EditArticleComponent /> component  1`] = `ShallowWrapper {}`;
